lib/healthcheck/manager.go

```diff
@@ -0,0 +1,371 @@
+			ResourceWatcherConfig: services.ResourceWatcherConfig{
+				Client:    m.cfg.Events,
+				Clock:     m.cfg.Clock,
+				Component: m.cfg.Component,
+				Logger:    m.logger,
+			},
+		},
+	)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	m.configWatcher = watcher
+	m.logger.DebugContext(ctx, "Started health check config resource watcher")
+	go func() {
+		defer watcher.Close()
+		defer m.logger.DebugContext(ctx, "Stopped health check config resource watcher")
+		for {
+			select {
+			case configs := <-watcher.ResourcesC:
+				m.updateConfigs(ctx, configs)
+			case <-watcher.Done():
+				return
+			}
+		}
+	}()
+	return nil
+}
+
+// updateConfigs updates the manager's known health check configs and
+// recalculates the matching config for each registered target resource.
+func (m *manager) updateConfigs(ctx context.Context, cs []*healthcheckconfigv1.HealthCheckConfig) {
+	// Config priority is by ascending order of name - the first config to match, wins.
+	slices.SortFunc(cs, func(a, b *healthcheckconfigv1.HealthCheckConfig) int {
+		return cmp.Compare(a.GetMetadata().GetName(), b.GetMetadata().GetName())
+	})
+	configs := make([]*healthCheckConfig, 0, len(cs))
+	for _, c := range cs {
+		configs = append(configs, newHealthCheckConfig(c))
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.configs = configs
+	m.updateWorkersLocked(ctx)
+}
+
+// startWorkerUpdater periodically evaluates and updates the matching health
+// check config for each worker.
+// The config for a worker may change if its target resource labels change
+// dynamically.
+func (m *manager) startWorkerUpdater(ctx context.Context) {
+	updateInterval := interval.New(interval.Config{
+		Duration: time.Minute,
+		Jitter:   retryutils.SeventhJitter,
+		Clock:    m.cfg.Clock,
+	})
+	go func() {
+		defer updateInterval.Stop()
+		for {
+			select {
+			case <-updateInterval.Next():
+				m.mu.Lock()
+				m.updateWorkersLocked(ctx)
+				m.mu.Unlock()
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+}
+
+// updateWorkersLocked evaluates and updates the matching health check config
+// for each worker.
+func (m *manager) updateWorkersLocked(ctx context.Context) {
+	for _, w := range m.workers {
+		newCfg := m.getConfigLocked(ctx, w.GetTargetResource())
+		w.UpdateHealthCheckConfig(newCfg)
+	}
+}
+
+// getConfigLocked gets a matching config for the the given resource or returns
+// nil if no config matches.
+func (m *manager) getConfigLocked(ctx context.Context, r types.ResourceWithLabels) *healthCheckConfig {
+	for _, cfg := range m.configs {
+		matched, _, err := services.CheckLabelsMatch(
+			types.Allow,
+			cfg.getLabelMatchers(r.GetKind()),
+			nil, // userTraits
+			r,
+			false, // debug
+		)
+		if err != nil {
+			m.logger.WarnContext(ctx, "Health check config failed to match",
+				"health_check_config", cfg.name,
+				"error", err,
+			)
+			continue
+		}
+		if matched {
+			return cfg
+		}
+	}
+	return nil
+}
+
+type testEvent struct {
```

Comment:
not a big fan of seeing "test" code in production code. Is it possible to avoid defining the `testEvent` struct? maybe something like
```
	// test event channels
	onCheckSuccess chan string
	onCheckFail chan string
	onConfigUpdate chan string
```

Or instead of calling it `testEvent`, maybe call it workerEvent or something.